### ISSUE TITLE
ABI issue with LLVM11 and D3D12 (__stdcall)

### ISSUE THREAD

**Issue by slomp** [halide/Halide#5156]
The following code snippet from `d3d12compute.cpp`:
```
D3D12_CPU_DESCRIPTOR_HANDLE hCPU = binder->CPU[UAV];
hCPU.ptr += i * descriptorSize;
(*device)->CreateUnorderedAccessView(NULL, NULL, &NullDescUAV, hCPU);
```
produces the following assembly
```
LLVM 10                                               LLVM 11
   0x2c619a06bf1:       mov    %r14,-0x1d0(%rbp)         0x19f89f06c01:       mov    %r14,-0x1e0(%rbp)
   ...                                                   ...
   0x2c619a06c09:       mov    %rcx,-0x220(%rbp)         0x19f89f06c19:       mov    %rcx,-0x1d0(%rbp)    <------
                                                         0x19f89f06c20:       lea    -0x1d0(%rbp),%rcx
   0x2c619a06c10:       mov    %rcx,0x20(%rsp)           0x19f89f06c27:       mov    %rcx,0x20(%rsp)
   ...                                                   ...
                        callq  *%rax                                          callq  *%rax
```
On the left, the value of `hCPU.ptr` is in `%rcx`, which is then placed on the stack (`0x20(%rsp)`) to serve as an argument to the subsequent `callq` to `CreateUnorderedAccessView`. This is the correct ABI behavior.

On the right, LLVM11 decided to replace the value of `%rcx` with the **address of that value**  (notice the `lea` instruction).
This is basically turning something that should have been passed by-value into something that is passed by-reference

In short, LLVM 11 is insisting in passing the struct by reference, even though it fits entirely on a 64bit word and should have been passed by-value:
```
typedef struct D3D12_CPU_DESCRIPTOR_HANDLE {
    SIZE_T ptr;
} D3D12_CPU_DESCRIPTOR_HANDLE;
```

**Comment by abadams** [comment]
Could this be a target triple issue?

**Comment by slomp** [comment]
We'll see, it's possible; I'm working on a clean, minimal reproducible code snippet for further investigation.

**Comment by slomp** [comment]
Update:
Yes, it looks like we're muffling a rather important warning:
```
warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes]
```
because `mini_d3d12.h` is imposing the following:
```
#pragma clang diagnostic ignored "-Wignored-attributes"
```
So I think this is indeed related to target triple (or quadruple/quintuple).

`le64` does not recognize `__stdcall`.

**Comment by slomp** [comment]
I think there's o way around: we'll have to adopt what has been proposed here:
https://github.com/halide/Halide/issues/3815

**Comment by slomp** [comment]
I can verify: building all modules with `x86_64-unknown-windows-unknown` worked with LLVM 11.
I also went ahead and disabled the d3d12 asm abi patches and can verify that they won't be needed if we use the proper triple/quadruple.

**Comment by slomp** [comment]
That said, the matter at hand is:
How to write the necessary CMakeLists.txt logic such that this triple is used only when compiling runtime modules to be used for windows (basically, what's the best way to detect we actually want to produce windows runtime modules)?
@shoaibkamil suggested generating the runtime modules for both combinations (`le64-unknown-unknown-unknown` and `x86_64-unknown-windows-unknown`) and let Halide deal with it at runtime module load time.

**Comment by abadams** [comment]
I don't think it's feasible to compile all the runtime modules both ways. Better to have a single windows runtime module hiding all the stdcall stuff in it.

**Comment by slomp** [comment]
Wouldn't that potentially cause something like this:
```
error: linking module flags 'wchar_size': IDs have conflicting values in 'd3d12compute_64_debug' and 'module_jit_ref_count_64_debug'
```

**Comment by abadams** [comment]
The target triple gets rewritten at initial module assembly time in LLVM_Runtime_Linker.cpp. The only reason that setting it in the build matters is because clang uses it for some things even before it hits llvm IR (e.g. struct layout). We should be able to set it as matching on the individual pieces before we ask llvm to link the modules together into a single runtime.

**Comment by slomp** [comment]
To summarize what needs to be done:

We need to add some sort of `d3d12-wrapper.cpp` which reroutes arguments from one ABI to another. This way, only `d3d12-wrapper.cpp` runtime module needs to be explicitly built with the `x86_64-unknown-windows-unknown` target. Ideally, we could even conceive the accompanying header `d3d12-wrapper.h` to follow the always trustworthy `extern C` ABI.

Based on @abadams comments above:
The start of `link_modules` in `LLVM_Runtime_Linker.cpp` rewrites all the target triples according to the Halide target before combining the modules, so the above strategy should work.

**Comment by slomp** [comment]
Ok, I have a proof-of-concept here:
https://github.com/halide/Halide/compare/slomp/d3d12-abi-fix

It's incomplete, but it gets the job done and it works.
We only build the `d3d12wrapper` using the required windows target triple.
One good thing is that I was able to remove the asm trampolines we used to rely on.
On the other hand, that this wrapper is tricky to "masquerade" as a D3D12 interface... Still beats the asm stubs, though.

**Comment by slomp** [comment]
With a little bit of cleanup, I think we could integrate this partial implementation as is, and add/polish more as needed in the future.

**Comment by slomp** [comment]
It's working on LLVM 10 and 11, but funny enough I am getting this when running `lesson_12_using_the_gpu.exe`:
```
Running pipeline on CPU:
error: linking module flags 'wchar_size': IDs have conflicting values in 'cache_64' and 'posix_allocator_64'
```
@abadams Any clues?

**Comment by abadams** [comment]
I'll take a look. One option is to just drop llvm 9 support one month early (11 should be released in a month or so, judging by the timing in previous years)

**Comment by slomp** [comment]
@abadams Actually, don't even bother, I made an egregious mistake that needs to be addressed first :(
https://github.com/halide/Halide/compare/slomp/d3d12-abi-fix#diff-ad0403fd84d3bdbcca569688ee153fdbR168

**Comment by slomp** [comment]
Yeah, after fixing the issue above, we're back with this on LLVM 10 and 11 too:
```
error: linking module flags 'wchar_size': IDs have conflicting values in 'cache_64' and 'posix_allocator_64'
```
I think it's because of some shenanigans where the wchar size differs in clang and Win32.

**Comment by abadams** [comment]
Can repro, and I think I see the fix.

Looking at the build, the windows_*_32 modules are already compiled with a target triple that supports stdcall. For me it worked to just rename d3d12*.cpp to windows_d3d12*.cpp, and then extend the existing build logic around targets that start with windows_ to include 64-bit too.

Also remember that you can't include system headers in runtime modules. Compiled fine for me to replace stdint.h with runtime_internal.h and just delete the includes in d3d12compute.cpp

**Comment by abadams** [comment]
See branch abadams/d3d12abi

**Comment by shoaibkamil** [comment]
@abadams's strategy works around the other issue I mentioned to @slomp offline: `windows_get_symbol.cpp` uses `wchar_t` directly, which will be defined differently in Windows targets than non-Windows.  `mini_d3d12.h` already works around this by defining `WCHAR` to always be a 16-bit type.  That's possibly the cause of seeing the particular issues with `wchar_size`.

**Comment by abadams** [comment]
For posterity, my "fix" was to just drop any module flags from modules not starting with windows_ when combining the initial modules. This only works if no modules outside of `windows_*` ones ever use wchar

**Comment by shoaibkamil** [comment]
Right: we should still make the change to `windows_get_symbol.cpp` so if any modules _do_ ever end up using wchar, it won't be an issue.

**Comment by abadams** [comment]
So we think windows_get_symbol has a pre-existing bug then for 64-bit windows? Why does it work if it thinks wchar size is 4?

Oh. The way it uses it doesn't depend on the size of it except for a sizeof, and the effect of that is to malloc 2x too much memory for something, which isn't visible. Apart from that it could equally be void *

**Comment by slomp** [comment]
I wonder if w should leave `d3d12wrapper.cpp` and `d3d12compute.cpp` as is, remove them from the list of modules, and then "#incude" them on their respective `windows_d3d12*.cpp` modules... This way, if we need special targets for Windows-on-ARM, we could reuse the code and stub it in `win_arm_d3d12*.cpp`.

**Comment by abadams** [comment]
We should just do exactly the same thing as with metal for uniformity and have trivial wrappers for both x86 and arm: (see metal_obj_arm.cpp vs metal_obj_x86.cpp). Is that what you're proposing?

So we'd have windows_x86_d312compute.cpp and windows_arm_d3d12compute.cpp, and they'd both include d3d13compute.cpp, which is not mentioned anywhere else in the build.

**Comment by slomp** [comment]
Yeap, pretty much, just a trivial wrapping.
Anyway, in that case, I think we don't even need d3d12wrapper, we can just stick to d3d12compute.
I'll give it a try on a clean slate

**Comment by slomp** [comment]
Ok, here's an cleaner branch with the new wrapping scheme:
https://github.com/halide/Halide/compare/slomp/d3d12-abi-patch

Unfortunately, the following error persists:
```
error: linking module flags 'wchar_size': IDs have conflicting values in 'windows_abort_64' and 'halide_buffer_t_64'
```

**Comment by abadams** [comment]
You need to pull my change to link_modules from my branch too

**Comment by slomp** [comment]
Yeah, I forgot that.
Anyway, now I am getting some sort of crash/early-exit at `lesson_12_using_the_gpu.cpp:109`:
```
curved.compile_jit(target);
```
This is actually the JITting of the CPU path of the tutorial.
@shoaibkamil  suspects that it could be crashing when constructing the initial module, but I can't break into Halide at the moment in my current setup.

**Comment by slomp** [comment]
The point of the crash seems to be in "JIT Compiling halide_buffer_copy". Need to dig deeper to know for sure.

**Comment by slomp** [comment]
More specifically, it's when `void *f = (void *)ee.getFunctionAddress("halide_buffer_copy");` gets called.

**Comment by abadams** [comment]
If you compile halide with -DHALIDE_ENABLE_EXCEPTIONS=OFF you get actual error messages on windows. Current hypothesis is that one of the things the target triple passed to clang controls is c++ name mangling. There's one module calling halide_abort assuming one type of name mangling, and the module that defines it starts with windows_, so it uses another type of name mangling.

**Comment by abadams** [comment]
Yes, that seems to be the issue. I'll make a PR that just avoids C++ manging issues for all cross-module dependencies

**Comment by abadams** [comment]
Patching in #5208 seems to work for me. It's crazy slow though - local laplacian is 200ms for d3d12compute and 2ms for cuda. Not sure what's going wrong.

**Comment by slomp** [comment]
Ah, great, many thanks for finding this.
I was almost going mad with stepping into LLVM in Debug mode...
I'll put the pieces together tomorrow and verify it works.
As for the slow down, yeah, I think there are a few too many hard host-device sync points in the d3d12 back-end that need to go away.

**Comment by shoaibkamil** [comment]
Awesome!  I could _not_ figure out how to get errors to show up correctly on Windows, so that's good to know.

> local laplacian is 200ms for d3d12compute and 2ms for cuda

Yikes, I guess I know what we have to work on next :)

**Comment by slomp** [comment]
So, I merged https://github.com/halide/Halide/pull/5208 into https://github.com/halide/Halide/compare/slomp/d3d12-abi-fix but we're still lost in mangling:
```
Internal Error at halide\src\JITModule.cpp:232 triggered by user code at :
Condition failed: result != 0:
HalideJITMemoryManager: unable to find address for _ZN6Halide7Runtime8Internal12halide_abortEv
```

**Comment by slomp** [comment]
It looks like that namespace mangling is still an issue, so maybe it's related to what @steven-johnson just mentioned in https://github.com/halide/Halide/pull/5208?

**Comment by abadams** [comment]
That specific error is the one the branch fixed for me. Try a full clean? (I had issues with this)

**Comment by slomp** [comment]
Yeah, I just made a clean build and that error went away.

However, now it crashes at this point:
```
JIT compiling shared runtime for x86-64-windows-avx-avx2-f16c-fma-jit-sse41-user_context
JIT Compiling halide_abort
```
I am not getting any messagesm, even with `-DHALIDE_ENABLE_EXCEPTIONS=OFF`

**Comment by abadams** [comment]
Sorry, the cmake variable changed. Check CMakeCache.txt, but I think it's Halide_ENABLE_EXCEPTIONS=OFF now

**Comment by slomp** [comment]
Alright, I can verify that it works now. Complete wipe-out and rebuild seems to have solved the issue I was having above.

**Comment by slomp** [comment]
Now that https://github.com/halide/Halide/pull/5208 has been merged, I'll resync with master and run a clean build on LLVM 10 and 11 to make sure everything is working before opening a PR for `slomp/d3d12-abi-fix`.
Thanks for spotting the linking issue, @abadams !

**Comment by slomp** [comment]
Tested on LLVM 11, 10 and 9. It's working fine!
PR here: https://github.com/halide/Halide/pull/5217

**Comment by slomp** [comment]
PR open: https://github.com/halide/Halide/pull/5217

**Comment by slomp** [comment]
PR #5217 merged.
